### ISSUE TITLE
Side effects setting download parts > 1

### ISSUE THREAD

**Issue by OtwakO** [erickutcher/httpdownloader#321]
Not sure if this is actually how it works, but it seems like as downloads are near complete, it still uses the amount of download parts set in the options. It would be wasting time firing connections to download very small portion of the download, to the point that firing connections takes longer than the actual transfer. Maybe provide a setting for minimum size threshhold for download parts, lets say if you set it to 50MB and download parts to 10, if there are 100MB left to download, it would only start 2 connections instead of 10.

**Comment by erickutcher** [comment]
Each connection will download whatever the size of the file is divided by the number of parts that were set when it was added. So If you have a file that's 500MB and it was set to use 10 parts, then each connection will download 50MB. There will only ever be 10 connections made for it unless you have "Reallocate parts to maximize connections" enabled. That setting will split any remaining parts to maintain the number of parts that was initially set (10 in this example).

There shouldn't be any wasted connections.

**Comment by OtwakO** [comment]
My bad, this might be a problem with specific file hosting site? in this case gofile. It seems like it's not assigning all download parts to the total size, instead it ends up downloading ~20% of whatever thats left with 10 connections and then firing another 10 connection for the next 20%, repeating until it's done, which take longer and longer when download are near 100%, firing 10 connections just to download >1mb. Taking 20s+ seconds on 99%+ complete on a rather small file because of this.

**Comment by erickutcher** [comment]
If the site is gofile.io, then it appears that the maximum number of simultaneous connections per download is 3. It could be per IP address, but I didn't test that. In any event, if you set it to 10, then 7 of the connections are going to fail and the remaining 3 will continue until their parts finish. HTTP Downloader will then attempt to retry the failed parts once the first 3 finish, but it will only make 1 connection for each of the remaining parts unless you have "Reallocate parts to maximize connections" enabled. If that's enabled then the 7 remaining parts will be reallocated into 10 and it'll end up using 10 total connections to finish downloading the file.

I would suggest adding `https://*.gofile.io` to the Site Manager and setting its parts to 3.

If you have any downloads in progress and want to lower the number of parts you can right click on it, select "Update Download..." and update the parts to an appropriate number.

Everything you've described sounds like normal behavior.

**Comment by OtwakO** [comment]
Yeah, seems like a restrain on gofile's side, specifically the store4.gofile.io domain from what I've tested, while domain like file10.gofile.io works just fine with 20 download parts. Probably best to add https://*.gofile.io to site manager. But I'm still quite curious on how other downloaders (tested IDM) counter this problem (works fine even on store4.gofile.io).

**Comment by erickutcher** [comment]
Try starting a download from gofile in HTTP Downloader and have it use as many connections as it can. Then download the same file in IDM at the same time. I suspect that gofile is limiting connections based on the IP address and if that's the case, then the IDM download will probably fail.

I don't think there's any counters or workarounds in other download managers. It's probably just coincidental.

**Comment by OtwakO** [comment]
Just reporting that it seems like gofile did some changes that messes with downloaders as of today, only the html is downloaded instead of the actual file. This is not an issue with this specific downloader but rather a PSA.
